src/main/scala/com/microsoft/hyperspace/index/rules/JoinIndexRuleV2.scala

```diff
@@ -0,0 +1,199 @@
+  def apply(plan: LogicalPlan): LogicalPlan = {
+    if (HyperspaceConf.joinV2RuleEnabled(spark)) {
+      plan transformUp {
+        case join @ Join(l, r, _, Some(condition)) if eligible(l, r, condition) =>
+          updatePlan(join)
+      }
+    } else {
+      plan
+    }
+  }
+
+  private def contains(
+      attributes: Set[AttributeReference],
+      attribute: AttributeReference): Boolean = {
+    attributes.exists(_.semanticEquals(attribute))
+  }
+
+  /**
+   * Return a logical plan with source data replaced by index data. The index should cover the
+   * relevant join columns as indexed columns. The index should also cover only the relevant
+   * columns from the list of allReferencedCols from this side of the plan.
+   *
+   * @param relation Source logical relation.
+   * @param joinCols All columns used in join predicate. This is a superset of columns from the
+   *                 relation as well as other join columns from other relations.
+   * @param allReferencedCols All columns referenced in the logical plan on any given side of Join
+   *                          node. This list of columns includes the output columns from the
+   *                          plan as well as other referenced columns from intermediate nodes.
+   *
+   * @return Logical plan with eligible indexes if found, otherwise return original relation.
+   */
+  private def setIndexes(
+      relation: LogicalRelation,
+      joinCols: Set[AttributeReference],
+      allReferencedCols: Seq[Attribute]): LogicalPlan = {
+    val indexManager = Hyperspace
+      .getContext(spark)
+      .indexCollectionManager
+
+    // TODO: the following check only considers indexes in ACTIVE state for usage. Update
+    //  the code to support indexes in transitioning states as well.
+    //  See https://github.com/microsoft/hyperspace/issues/65.
+    val allIndexes = indexManager.getIndexes(Seq(Constants.States.ACTIVE))
+
+    val allCols = (joinCols ++ allReferencedCols.toSet).map(_.asInstanceOf[AttributeReference])
+
+    val allReqdCols =
+      relation.outputSet
+        .filter(c => contains(allCols, c.asInstanceOf[AttributeReference]))
+        .map(_.name)
+    val reqdIdxCols = relation.outputSet
+      .filter(
+        c =>
+          contains(
+            joinCols.map((col: Attribute) => col.asInstanceOf[AttributeReference]),
+            c.asInstanceOf[AttributeReference]))
+      .map(_.name)
+
+    val usableIndexes = allIndexes
+      .filter { index =>
+        index.config.indexedColumns.toSet.equals(reqdIdxCols.toSet) &&
+        allReqdCols.forall(
+          (index.config.indexedColumns ++ index.config.includedColumns).contains(_))
+      }
+
+    if (usableIndexes.isEmpty) {
+      return relation
+    }
+
+    val candidateIndexes = RuleUtils.getCandidateIndexes(spark, usableIndexes, relation)
+    if (candidateIndexes.isEmpty) {
+      return relation
+    }
+
+    RuleUtils.transformPlanToUseIndex(spark, candidateIndexes.head, relation, true)
```

Comment:
Not using return is one of best practices in Scala language. I'd like to hear other opinions on this @imback82 @pirz.

Reply:
I am fine with using `return` if it "greatly" improves the readability.

In this case, I am fine with not using `return` since non-`return` version is clear enough.

Reply:
fair enough. Please have a look at the updated code